### PR TITLE
chore: Bubble precise Usages instead of Locations

### PR DESCRIPTION
**Pull request by varungandhi-src**
This patch changes the code so that the core APIs we need
for precise usagesForSymbol actually return (Upload)Usages instead
of just (Upload)Locations, so that we have the usage kind information
as well as the symbol names available. I will be handling the GraphQL layer
for converting the Usages to appropriate Resolvers in a follow-up PR.

## Test plan

Modified a bunch of existing mock tests. 🤷🏽

### CODE REVIEW

File: internal/codeintel/codenav/service_new_test.go [hunk 1]
```diff
@@ -56,20 +58,25 @@ func TestGetDefinitions(t *testing.T) {
 		}
 		mockRequestState.SetUploadsDataLoader(uploads)
 
-		locations := []shared.Location{
-			{UploadID: 51, Path: uploadRelPath("a.go"), Range: testRange1},
-			{UploadID: 51, Path: uploadRelPath("b.go"), Range: testRange2},
-			{UploadID: 51, Path: uploadRelPath("a.go"), Range: testRange3},
-			{UploadID: 51, Path: uploadRelPath("b.go"), Range: testRange4},
-			{UploadID: 51, Path: uploadRelPath("c.go"), Range: testRange5},
+		locations := []shared.UsageBuilder{
```
Comment (varungandhi-src) [inline internal/codeintel/codenav/service_new_test.go:61]: Code here needs more changes

File: internal/codeintel/codenav/transport/graphql/root_resolver_usages.go [hunk 2]
```diff
@@ -38,6 +38,35 @@ type usageResolver struct {
 
 var _ resolverstubs.UsageResolver = &usageResolver{}
 
+//func NewPreciseUsageResolver(usage shared.UploadUsage) resolverstubs.UsageResolver {
```
Comment (varungandhi-src) [inline internal/codeintel/codenav/transport/graphql/root_resolver_usages.go:41]: TODO: Remove this

File: internal/codeintel/codenav/transport/graphql/root_resolver.go [hunk 3]
```diff
@@ -319,6 +320,76 @@ func (r *rootResolver) MakeGitTreeTranslator(repo *sgtypes.Repo) codenav.GitTree
 	return codenav.NewGitTreeTranslator(r.gitserverClient, *repo)
 }
 
+//// Pre-condition: 'args' must've been normalized before
```
Comment (varungandhi-src) [inline internal/codeintel/codenav/transport/graphql/root_resolver.go:323]: TODO: Remove this

File: internal/codeintel/codenav/shared/types.go [hunk 4]
```diff
@@ -86,25 +89,43 @@ func (k UsageKind) String() string {
...
-	Range  scip.Range
-	Symbol string
-	// SymbolRoles represents the 'role' of the underlying occurrence
-	// which was used to construct this UsageBuilder.
-	//
-	// TODO: Make this field private after removing extractLocationsFromPosition
-	SymbolRoles scip.SymbolRole
+	Range       scip.Range
+	Symbol      string
+	symbolRoles scip.SymbolRole
```
Comment (varungandhi-src) [inline internal/codeintel/codenav/shared/types.go:94]: TODO: Keep the doc comment

File: internal/codeintel/codenav/shared/types.go [hunk 5]
```diff
@@ -43,10 +46,10 @@ func (u Usage) ToLocation() Location {
...
 
 const (
-	UsageKindDefinition UsageKind = iota
-	UsageKindReference
-	UsageKindImplementation
-	UsageKindSuper
+	UsageKindDefinition     UsageKind = 1
+	UsageKindReference      UsageKind = 2
+	UsageKindImplementation UsageKind = 3
+	UsageKindSuper          UsageKind = 4
```
Comment (kritzcreek) [inline internal/codeintel/codenav/shared/types.go:52]: What's this change about? Do these numbers need to line up with some protobuf generated schema? If so we should add a comment.

File: internal/codeintel/codenav/shared/types.go [hunk 6]
```diff
@@ -141,6 +164,48 @@ type UploadLocation struct {
...
+		}
+		if c := cmp.Compare(loc1.Path.RawValue(), loc2.Path.RawValue()); c != 0 {
+			return c
+		}
+		// This order is reversed as we want to prefer newer uploads over older ones
+		// for same (repo, path), see Filter below
+		if c := cmp.Compare(loc2.Upload.ID, loc1.Upload.ID); c != 0 {
+			return c
+		}
+		return loc1.TargetRange.ToSCIPRange().Compare(loc2.TargetRange.ToSCIPRange())
```
Comment (kritzcreek) [inline internal/codeintel/codenav/shared/types.go:197]: `CompareStrict`?

**Review by kritzcreek — CHANGES_REQUESTED** [review]
Just some things I spotted, will work through the rest in a bit

File: internal/codeintel/codenav/shared/types.go (hunk 5 above)
Comment (varungandhi-src) [reply internal/codeintel/codenav/shared/types.go:52]: Don't want to accidentally use zero values

File: internal/codeintel/codenav/service.go [hunk 7]
```diff
@@ -580,11 +597,13 @@ func (s *Service) getCodeIntelligenceRange(ctx context.Context, args RequestArgs
...
+	_, _, _ = definitions, references, implementations
+
 	return AdjustedCodeIntelligenceRange{
 		Range:           adjustedRange,
-		Definitions:     definitions,
-		References:      references,
-		Implementations: implementations,
+		Definitions:     nil, // TODO: Do slice conversion
+		References:      nil,
+		Implementations: nil,
```
[the remark this answers is not in the document]
Comment (varungandhi-src) [reply internal/codeintel/codenav/service.go]: Fixed

File: internal/codeintel/codenav/internal/lsifstore/document_metadata_test.go [hunk 8]
```diff
@@ -132,29 +133,63 @@ func TestGetRanges(t *testing.T) {
 		valueHoverText    = "```ts\n(parameter) value: T | T[] | null | undefined\n```\nThe value to test."
 		tHoverText        = "```ts\nT: T\n```"
 	)
-
+	const (
+		nonEmptySymbol = "scip-typescript npm template 0.0.0-DEVELOPMENT src/util/`helpers.ts`/nonEmpty()."
+		tSymbol        = "scip-typescript npm template 0.0.0-DEVELOPMENT src/util/`helpers.ts`/nonEmpty().[T]"
+		valueSymbol    = "scip-typescript npm template 0.0.0-DEVELOPMENT src/util/`helpers.ts`/nonEmpty().(value)"
```
[the remark this answers is not in the document]
Comment (varungandhi-src) [reply internal/codeintel/codenav/internal/lsifstore/document_metadata_test.go:139]: No, these are hard-coded based on the existing index, this patch is just moving code around. This function starts with `store := populateTestStore(t)`

[the remark this answers is not in the document]
Comment (kritzcreek) [reply internal/codeintel/codenav/internal/lsifstore/document_metadata_test.go:139]: I hate that testStore so much...

**Review by kritzcreek — APPROVED** [review]
Nice!

**Pull request merged**